doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,84 @@
+that user can tell some other user about its results,
+and the second user is guaranteed to observe those results (unless they've changed in the meantime).
+
+However, we cannot have a single timestamp number that we guarantee always goes up, because timestamps do not always have the same meaning.
+We have solved the different meaning problem with timelines, which allow joining different sources only if their timestamp number means the same thing.
+For linearizability, we need to have a timestamp number for each timeline.
+
+Add a `HashMap<Timeline, Timestamp>` to track the highest timestamp for a given timeline.
+`sequence_peek`, `sequence_tail`, and `sequence_end_transaction` are the three places timestamps are generated for transactions.
+These three functions will call some new helper function that will either advance or return the most recent timestamp in the HashMap.
+
+This linearizability system can thus be viewed as a layer above the current timestamp determination functions, timelines, and time domains, since it specifically only cares about timestamps that users initiated.
+
+### System Clock
+
+The system clock (calls to `SystemTime::now()`) contributes to timestamps for realtime sources, table reads, and table writes.
+It may decrease (due to, say, some time synchronization program like NTP setting the clock).
+Linearizability is preserved in this case with the above `HashMap`.
+The effect would be a negative user experience until the system clock catches up with what it once was.
+
+It would result in no new data will be produced from (rather, we will not advance the frontiers of) realtime sources.
+SELECTs will continue to function normally, but without any updates until the system clock catches up to what it was.
+Tables also have a side effect that if reads and writes are interleaved, the writes will increase the timestamp by at least 1 (if the system clock did not increase).
+Thus in worst-case scenarios table writes would be observed, but no new data from realtime sources, until the system clock catches up.
+
+## Alternatives
+
+Instead of advancing timestamps on each transaction, we use index `upper`s to advance the timestamp-per-timeline.
+When `update_upper` is run, it would bump the timeline's timestamp to the max of what is there and the new upper.
+This would potentially cause read queries to block if another index in the same timeline is noticably ahead of another.
+
+## Open questions
+
+Will this cause regular blocking for realtime sources if tables are involved a lot because those always bump the ts, but realtime sources might take up to 1s to catch up to that?
+
+Will the change to use the durability frontier instead of the upper also cause up to 1s pauses because it is only persisted to disk each 1s?
+If so, this seems like a thing we could optimize for: write to disk more frequently, and measure that it doesn't get too slow.
+
+What are the benchmarking diffs after this change?
+Does it affect time to first row?
```

Comment:
I think a related thing to nail down is what degrees of freedom do we have to recover performance. If linearizability is performance limiting, that is good to know (imo it is more important) but we want to be aware of the degrees of freedom.